Remove doc comments from concrete floating-point types

stephentyrone opened the pull request:
... when they are already present on the protocol. I don't *think* that we need them anymore for xcode documentation purposes. There are reasonable arguments both ways on this:
1. when you're editing the concrete implementations, it's sometimes nice to have the doc comment right there.
2. but it needlessly repetitive, and introduces the opportunity for comments to get out of sync.
3. it also adds noise; it would be nice for information density if the implementation only had implementation notes.

Also includes some minor formatting fixups, and a bugfix for isCanonical on platforms that do not support subnormals (armv7, currently, which is sufficiently minor that I'm OK with stealth fixing it like this).

stephentyrone committed:
Remove doc comments from concrete floating-point types

... when they are already present on the protocol. I don't *think* that we need them anymore for xcode documentation purposes. There are reasonable arguments both ways on this:
1. when you're editing the concrete implementations, it's sometimes nice to have the doc comment right there.
2. but it needlessly repetitive, and introduces the opportunity for comments to get out of sync.
3. it also adds noise; it would be nice for information density if the implementation only had implementation notes.

stephentyrone commented:
@swift-ci please test

xwu commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -231,89 +196,60 @@ extension ${Self}: BinaryFloatingPoint {
+    // All Float and Double encodings are canonical in IEEE 754.
+    //
+    // On platforms that do not support subnormals, we treat them as
+    // non-canonical encodings of zero.
```
I think we should document this in the corresponding protocol requirement in `FloatingPoint.swift`. I like LLVM's description that offers examples:

> Examples of non-canonical encodings:
>
> * x87 pseudo denormals, pseudo NaNs, pseudo Infinity, Unnormals. These are converted to a canonical representation per hardware-specific protocol.
> * Many normal decimal floating-point numbers have non-canonical alternative encodings.
> * Some machines, like GPUs or ARMv7 NEON, do not support subnormal values. These are treated as non-canonical encodings of zero[...]

stephentyrone replied:
Done!

xwu commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -883,53 +600,30 @@ extension ${Self}: BinaryFloatingPoint {
     self = result
   }
 
-  /// Creates a NaN ("not a number") value with the specified payload.
```
This isn't a protocol requirement, so we can't delete the documentation here.

stephentyrone replied:
Restored.

stephentyrone committed:
Add examples to documentation for FloatingPoint.isCanonical

In particular, document that subnormal encodings are treated as non-canonical zeros on platforms that flush to zero.

stephentyrone committed:
Restore doc comment on concrete fp types init(payload: signaling:)

This one isn't a protocol requirement, so I was slightly overzealous removing it.

stephentyrone commented:
@swift-ci please smoke test

stephentyrone commented:
@natecook1000 is on vacation this week, I'm going to wait to merge this until he has a chance to give it at least a cursory approval.

natecook1000 reviewed (commented):
Looks great, with some minor nits! I’m especially glad to see the repeated docs dropping off the extension methods. Do we have tests that show that the docs are being correctly inherited from the protocols down onto the concrete symbols?

stephentyrone committed:
Update stdlib/public/core/FloatingPoint.swift

Co-Authored-By: Nate Cook <natecook@apple.com>

stephentyrone committed:
Update stdlib/public/core/FloatingPoint.swift

Co-Authored-By: Nate Cook <natecook@apple.com>

stephentyrone commented:
> Do we have tests that show that the docs are being correctly inherited from the protocols down onto the concrete symbols?

I don't think this is testable for Swift, since it's an Xcode thing, but it seems to work robustly in 11.

stephentyrone commented on stdlib/public/core/FloatingPoint.swift:
```diff
@@ -1197,13 +1197,27 @@ public protocol FloatingPoint: SignedNumeric, Strideable, Hashable
+  ///   interpreted as non-canonical encodings of subnormal values.
+  ///
+  /// - Decimal floating-point admits a large number of non-canonical
+  ///   encodings. Consult the IEEE 754 standard for additional details.
```
Will do.

stephentyrone committed:
Change suggested by reviewer.

natecook1000 commented:
Interesting, I hadn’t noticed that!

stephentyrone commented:
@swift-ci please smoke test
